compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,410 @@
+
+        // Put the upper bound classes into the class hierarchy and count how many subclasses are there for each encountered class
+        val subClassCounts: MutableMap<ConcreteClass, AtomicInteger> = mutableMapOf()
+        val queue = ArrayDeque(upperBoundConcreteClassToTypes.keys)
+        while (queue.isNotEmpty()) {
+            for (superConcreteClass in queue.removeFirst().getSuperConcreteClasses(this)) {
+                subClassCounts.computeIfAbsent(superConcreteClass) {
+                    queue.addLast(it)
+                    AtomicInteger()
+                }.incrementAndGet()
+            }
+        }
+
+        val leafClassesOrInterfaces = upperBoundConcreteClassToTypes.keys.filter { subClassCounts[it]?.get() ?: 0 == 0 }
+        val leafClasses = leafClassesOrInterfaces.filter { !it.isInterface() }
+
+        if (mimicOldFrontendBehavior) {
+            // Old frontend allows multiple related classes!
+            val finalClassCount = leafClasses.count { it.isFinal() }
+            val enumClassCount = leafClasses.count { it.isEnum() }
+
+            if (finalClassCount >= 2 || enumClassCount >= 1 && leafClassesOrInterfaces.size >= 2) {
+                // FE1.0 allows unrelated non-final classes in org.jetbrains.kotlin.types.TypeIntersector. But, it doesn't allow
+                // * multiple unrelated final classes
+                // * an enum class with an unrelated class or interface
+                return false
+            }
+            if (finalClassCount == 0 && enumClassCount == 0) {
+                // FE1.0 simply bails out when it has determined all classes in question are open. See https://github.com/JetBrains/kotlin/blob/fe78e153f760be975cb817a068f1eb65160dec7f/compiler/frontend/src/org/jetbrains/kotlin/types/TypeIntersector.java#L79
+                return true
+            }
+        } else {
+            // Java force single inheritance, so any pair of unrelated classes are incompatible.
+            if (leafClasses.size >= 2) {
+                return false
+            }
+        }
+
+        // One final class and some other unrelated interface are not compatible
+        if (leafClasses.firstOrNull()?.isFinal() == true && leafClassesOrInterfaces.size > leafClasses.size) {
+            return false
+        }
+
+        // Check if the range formed by upper bounds and lower bounds is empty.
+        if (!lowerBounds.all { lowerBoundType ->
+                val classesSatisfyingLowerBounds =
+                    lowerBoundType.toConcreteClass(this)?.getThisAndAllSuperConcreteClasses(this) ?: emptySet()
+                leafClassesOrInterfaces.all { it in classesSatisfyingLowerBounds }
+            }
+        ) {
+            return false
+        }
+
+        if (upperBounds.size < 2) return true
+
+        // Base types are compatible. Now we check type parameters.
+
+        // Filter out integer literal types since they don't contain any type parameters.
```

Comment:
It's irrelevant now. But if you are curious, it was needed for checking base classes. It's removed here since there is no need to consider integer literal type when checking type parameters.